File: expense-tracker/app/src/main/java/com/farhan/paisatrack/ui/screens/ReportsScreen.kt
package com.farhan.paisatrack.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.farhan.paisatrack.data.TxnType
import com.farhan.paisatrack.ui.MainViewModel
import com.farhan.paisatrack.ui.components.SectionCard
import com.farhan.paisatrack.ui.theme.ExpenseRed
import com.farhan.paisatrack.ui.theme.IncomeGreen
import com.farhan.paisatrack.util.Format
import com.farhan.paisatrack.util.IconMap

@Composable
fun ReportsScreen(vm: MainViewModel) {
    val txns by vm.confirmedTxns.collectAsStateWithLifecycle()
    val categories by vm.allCategories.collectAsStateWithLifecycle()
    val accounts by vm.accounts.collectAsStateWithLifecycle()
    val catById = categories.associateBy { it.id }
    val accById = accounts.associateBy { it.id }

    val months = remember(txns) {
        txns.map { Format.monthKey(it.dateTime) to Format.month(it.dateTime) }
            .distinct()
            .sortedByDescending { it.first }
    }
    var selectedMonth by remember(months) { mutableStateOf(months.firstOrNull()?.first) }

    val monthTxns = txns.filter { selectedMonth == null || Format.monthKey(it.dateTime) == selectedMonth }
    val expense = monthTxns.filter { it.type == TxnType.EXPENSE }
    val income = monthTxns.filter { it.type == TxnType.INCOME }
    val totalExpense = expense.sumOf { it.amount }
    val totalIncome = income.sumOf { it.amount }

    val byCategory = expense.groupBy { it.categoryId }
        .map { (cid, list) -> cid to list.sumOf { it.amount } }
        .sortedByDescending { it.second }

    val byAccount = expense.groupBy { it.accountId }
        .map { (aid, list) -> aid to list.sumOf { it.amount } }
        .sortedByDescending { it.second }

    Column(
        Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp)
    ) {
        Text("Reports", style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.Bold)

        if (months.isNotEmpty()) {
            LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                items(months) { (key, label) ->
                    FilterChip(
                        selected = selectedMonth == key,
                        onClick = { selectedMonth = key },
                        label = { Text(label) }
                    )
                }
            }
        }

        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            BigStat(Modifier.weight(1f), "Income", Format.money(totalIncome), IncomeGreen)
            BigStat(Modifier.weight(1f), "Expense", Format.money(totalExpense), ExpenseRed)
        }

        SectionCard {
            Text("Net this period", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(6.dp))
            val net = totalIncome - totalExpense
            Text(
                Format.money(net),
                fontWeight = FontWeight.Bold,
                fontSize = 26.sp,
                color = if (net >= 0) IncomeGreen else ExpenseRed
            )
        }

        if (byCategory.isNotEmpty() && totalExpense > 0) {
            SectionCard {
                Text("Spending by category", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
                Spacer(Modifier.height(12.dp))
                byCategory.forEach { (cid, amount) ->
                    val cat = cid?.let { catById[it] }
                    val frac = (amount / totalExpense).toFloat()
                    BreakdownBar(
                        label = cat?.name ?: "Uncategorized",
                        amount = amount,
                        fraction = frac,
                        color = IconMap.parseColor(cat?.colorHex ?: "#636E72")
                    )
                    Spacer(Modifier.height(10.dp))
                }
            }
        }

        if (byAccount.isNotEmpty() && totalExpense > 0) {
            SectionCard {
                Text("Spending by account", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
                Spacer(Modifier.height(12.dp))
                byAccount.forEach { (aid, amount) ->
                    val acc = accById[aid]
                    val frac = (amount / totalExpense).toFloat()
                    BreakdownBar(
                        label = acc?.name ?: "Unknown",
                        amount = amount,
                        fraction = frac,
                        color = IconMap.parseColor(acc?.colorHex ?: "#6C5CE7")
                    )
                    Spacer(Modifier.height(10.dp))
                }
            }
        }

        if (txns.isEmpty()) {
            SectionCard {
                Text(
                    "Add some transactions to see your reports here.",
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
        Spacer(Modifier.height(80.dp))
    }
}

@Composable
private fun BigStat(modifier: Modifier, label: String, value: String, color: Color) {
    androidx.compose.material3.Card(
        modifier = modifier,
        shape = RoundedCornerShape(18.dp),
        colors = androidx.compose.material3.CardDefaults.cardColors(containerColor = color.copy(alpha = 0.12f))
    ) {
        Column(Modifier.padding(16.dp)) {
            Text(label, fontSize = 13.sp)
            Text(value, fontWeight = FontWeight.Bold, fontSize = 20.sp, color = color)
        }
    }
}

@Composable
private fun BreakdownBar(label: String, amount: Double, fraction: Float, color: Color) {
    Column {
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Text(label, fontWeight = FontWeight.Medium, fontSize = 14.sp)
            Text(
                "${Format.money(amount)}  ·  ${(fraction * 100).toInt()}%",
                fontSize = 13.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        Spacer(Modifier.height(6.dp))
        Box(
            Modifier
                .fillMaxWidth()
                .height(10.dp)
                .clip(RoundedCornerShape(6.dp))
                .background(MaterialTheme.colorScheme.surfaceVariant)
        ) {
            Box(
                Modifier
                    .fillMaxWidth(fraction.coerceIn(0.02f, 1f))
                    .height(10.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .background(color)
            )
        }
    }
}

File: farhans-wallet-flutter/lib/features/reports/reports_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/theme/tokens.dart';
import '../../core/utils/format.dart';
import '../../core/utils/icon_map.dart';
import '../../core/widgets/glass.dart';
import '../../data/models.dart';
import '../../state/wallet_state.dart';

class ReportsScreen extends ConsumerStatefulWidget {
  const ReportsScreen({super.key});

  @override
  ConsumerState<ReportsScreen> createState() => _ReportsScreenState();
}

class _ReportsScreenState extends ConsumerState<ReportsScreen> {
  String? _month;

  @override
  Widget build(BuildContext context) {
    final t = context.tokens;
    final async = ref.watch(walletProvider);
    return async.when(
      loading: () => const Center(child: CircularProgressIndicator()),
      error: (e, _) => Center(child: Text('Error: $e')),
      data: (data) {
        final months = <String, String>{};
        for (final tx in data.confirmed) {
          months[Dates.monthKey(tx.dateTime)] = Dates.month(tx.dateTime);
        }
        final monthKeys = months.keys.toList()..sort((a, b) => b.compareTo(a));
        _month ??= monthKeys.isNotEmpty ? monthKeys.first : null;

        final monthTxns = data.confirmed
            .where((tx) => _month == null || Dates.monthKey(tx.dateTime) == _month)
            .toList();
        final expense = monthTxns.where((tx) => tx.type == TxnType.expense).toList();
        final income = monthTxns.where((tx) => tx.type == TxnType.income).toList();
        final totalExpense = expense.fold<double>(0, (s, tx) => s + tx.amount);
        final totalIncome = income.fold<double>(0, (s, tx) => s + tx.amount);

        final byCat = <int?, double>{};
        for (final tx in expense) {
          byCat[tx.categoryId] = (byCat[tx.categoryId] ?? 0) + tx.amount;
        }
        final catEntries = byCat.entries.toList()..sort((a, b) => b.value.compareTo(a.value));

        final byAcc = <int, double>{};
        for (final tx in expense) {
          byAcc[tx.accountId] = (byAcc[tx.accountId] ?? 0) + tx.amount;
        }
        final accEntries = byAcc.entries.toList()..sort((a, b) => b.value.compareTo(a.value));

        return ListView(
          padding: const EdgeInsets.fromLTRB(Insets.lg, Insets.md, Insets.lg, 120),
          children: [
            Text('Reports', style: context.text.headlineMedium),
            const SizedBox(height: Insets.md),
            if (monthKeys.isNotEmpty)
              SizedBox(
                height: 38,
                child: ListView(
                  scrollDirection: Axis.horizontal,
                  children: monthKeys
                      .map((k) => Padding(
                            padding: const EdgeInsets.only(right: Insets.xs),
                            child: _chip(months[k]!, _month == k, () => setState(() => _month = k)),
                          ))
                      .toList(),
                ),
              ),
            const SizedBox(height: Insets.md),
            Row(
              children: [
                Expanded(child: _bigStat('Income', totalIncome, t.success)),
                const SizedBox(width: Insets.sm),
                Expanded(child: _bigStat('Expense', totalExpense, t.danger)),
              ],
            ),
            const SizedBox(height: Insets.md),
            GlassCard(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text('Net this period', style: context.text.titleMedium),
                  const SizedBox(height: 6),
                  Builder(builder: (_) {
                    final net = totalIncome - totalExpense;
                    return Text(Money.format(net),
                        style: context.text.headlineMedium
                            ?.copyWith(color: net >= 0 ? t.success : t.danger));
                  }),
                ],
              ),
            ),
            if (catEntries.isNotEmpty && totalExpense > 0) ...[
              const SizedBox(height: Insets.md),
              GlassCard(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Spending by category', style: context.text.titleMedium),
                    const SizedBox(height: Insets.md),
                    ...catEntries.map((e) {
                      final cat = e.key != null ? data.categoryById[e.key] : null;
                      return _bar(
                        cat?.name ?? 'Uncategorized',
                        e.value,
                        e.value / totalExpense,
                        AppIcons.parseColor(cat?.colorHex ?? '#636E72'),
                      );
                    }),
                  ],
                ),
              ),
            ],
            if (accEntries.isNotEmpty && totalExpense > 0) ...[
              const SizedBox(height: Insets.md),
              GlassCard(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Spending by account', style: context.text.titleMedium),
                    const SizedBox(height: Insets.md),
                    ...accEntries.map((e) {
                      final acc = data.accountById[e.key];
                      return _bar(
                        acc?.name ?? 'Unknown',
                        e.value,
                        e.value / totalExpense,
                        AppIcons.parseColor(acc?.colorHex ?? '#6C5CE7'),
                      );
                    }),
                  ],
                ),
              ),
            ],
            if (data.confirmed.isEmpty)
              Padding(
                padding: const EdgeInsets.only(top: 40),
                child: Text('Add some transactions to see your reports here.',
                    style: context.text.bodyMedium?.copyWith(color: t.textMid),
                    textAlign: TextAlign.center),
              ),
          ],
        );
      },
    );
  }

  Widget _bigStat(String label, double value, Color color) {
    return GlassCard(
      gradient: LinearGradient(
        colors: [color.withValues(alpha: 0.2), color.withValues(alpha: 0.05)],
        begin: Alignment.topLeft,
        end: Alignment.bottomRight,
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(label, style: context.text.bodySmall),
          const SizedBox(height: 4),
          Text(Money.format(value),
              style: context.text.titleLarge?.copyWith(color: color, fontWeight: FontWeight.w800)),
        ],
      ),
    );
  }

  Widget _bar(String label, double amount, double fraction, Color color) {
    final t = context.tokens;
    return Padding(
      padding: const EdgeInsets.only(bottom: Insets.sm),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text(label, style: context.text.bodyMedium?.copyWith(fontWeight: FontWeight.w600)),
              Text('${Money.format(amount)} · ${(fraction * 100).round()}%',
                  style: context.text.bodySmall?.copyWith(color: t.textMid)),
            ],
          ),
          const SizedBox(height: 6),
          ClipRRect(
            borderRadius: BorderRadius.circular(Corners.pill),
            child: Stack(
              children: [
                Container(height: 10, color: t.glassFill),
                TweenAnimationBuilder<double>(
                  tween: Tween(begin: 0, end: fraction.clamp(0.03, 1)),
                  duration: Motion.slow,
                  curve: Curves.easeOutCubic,
                  builder: (context, v, _) => FractionallySizedBox(
                    widthFactor: v,
                    child: Container(
                      height: 10,
                      decoration: BoxDecoration(
                        gradient: LinearGradient(colors: [color.withValues(alpha: 0.7), color]),
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _chip(String label, bool selected, VoidCallback onTap) {
    final t = context.tokens;
    return Pressable(
      onTap: onTap,
      child: AnimatedContainer(
        duration: Motion.fast,
        padding: const EdgeInsets.symmetric(horizontal: Insets.md, vertical: 8),
        decoration: BoxDecoration(
          color: selected ? t.accentA.withValues(alpha: 0.2) : t.glassFill,
          borderRadius: BorderRadius.circular(Corners.pill),
          border: Border.all(color: selected ? t.accentA.withValues(alpha: 0.5) : t.glassBorder),
        ),
        child: Text(label,
            style: TextStyle(
                color: selected ? t.accentA : t.textMid, fontWeight: FontWeight.w600, fontSize: 13)),
      ),
    );
  }
}
